packages/client/src/client.ts

```diff
@@ -0,0 +1,303 @@
+   * connection to this provider, this promise should successfully resolve.
+   */
+  fetchManifest(): Promise<PenumbraManifest> | undefined {
+    return this.config?.requestManifest;
+  }
+
+  /** Return the `Transport` representing the present connection, or create it if
+   * necessary by calling `connect` to obtain a `MessagePort` from the
+   * associated provider. An unconfigured client may be configured at this
+   * moment with the `providerOrigin` string parameter. */
+  public async connect(providerOrigin?: string): Promise<void> {
+    if (providerOrigin) {
+      this.config ??= this.createConfig(providerOrigin);
+    }
+    const { requestManifest, provider } = this.assertConfig(providerOrigin);
+    await requestManifest;
+    await provider.connect();
+    this.assertConnection();
+  }
+
+  /** Assert an active connection, and potentially init this client's transport. */
+  private assertConnection() {
+    assertProviderConnected(this.config?.origin);
+    this.connection ??= this.createConnection();
+    return this.connection;
+  }
+
+  /** Destroy any active connection and clients. */
+  private destroyConnection() {
+    void this.connection?.port.then(port => {
+      port.postMessage(false);
+      port.close();
+    });
+    this.connection = undefined;
+    this.serviceClients.clear();
+  }
+
+  /** Destroy any present connection, and call `disconnect` on the associated
+   * provider to release connection approval. If there is no present
+   * configuration, returns `undefined`. */
+  public disconnect(): Promise<void> | undefined {
+    const request = this.config?.provider.disconnect();
+    this.destroyConnection();
+    return request;
+  }
+
+  private connectService<T extends ServiceType>(service: T) {
+    const { transport } = this.assertConnection();
+    return createPromiseClient(service, transport);
+  }
+
+  /** Return a `PromiseClient<T>` for some `T extends ServiceType`, using this
+   * client's internal `Transport`. If you call this method before
+   * configuration, this method will throw.
+   *
+   * You should prefer to call this method *after* this client's connection has
+   * succeeded.
+   *
+   * If you call this method before a connection is resolved, a connection will
+   * be initiated but will not be awaited (as this is a synchronous method). If
+   * a connection initated this way is rejected, or does not resolve within the
+   * `defaultTimeoutMs` of this client's `transportOptions`, requests made with
+   * the returned `PromiseClient<T>` will throw.
+   */
+  public service<T extends ServiceType>(service: T) {
+    return (
+      (this.serviceClients.get(service.typeName) as PromiseClient<T> | undefined) ??
+      (this.serviceClients
+        .set(service.typeName, this.connectService(service))
+        .get(service.typeName) as PromiseClient<T>)
+    );
+  }
+
+  get addEventListener(): PenumbraProvider['addEventListener'] | undefined {
```

Comment:
todo: add JSDoc for these and the rest of the functions.

JSDocs should be detailed for each of the function, even if it just re-exports the injected provider's methods. This way, when users create the `client`, they will hover over the method (or choose it in the IDE's autocomplete) with the comments.

Reply:
just removing the forwarded provider method here as
1. the callback provides the functionality
2. the provider is accessible from the client if desired